libraries/microsoft-agents-authentication-msal/microsoft_agents/authentication/msal/msal_auth.py

```diff
@@ -186,3 +194,127 @@ def _resolve_scopes_list(self, instance_url: URI, scopes=None) -> list[str]:
+
+            if auth_result_payload:
+                return auth_result_payload.get("access_token")
+
+        return None
+
+    async def get_agentic_instance_token(
+        self, agent_app_instance_id: str
+    ) -> tuple[str, str]:
+        """Gets the agentic instance token for the given agent application instance ID.
+
+        :param agent_app_instance_id: The agent application instance ID.
+        :type agent_app_instance_id: str
+        :return: A tuple containing the agentic instance token and the agent application token.
+        :rtype: tuple[str, str]
+        """
+
+        if not agent_app_instance_id:
+            raise ValueError("Agent application instance Id must be provided.")
+
+        agent_token_result = await self.get_agentic_application_token(
+            agent_app_instance_id
+        )
+
+        authority = (
+            f"https://login.microsoftonline.com/{self._msal_configuration.TENANT_ID}"
+        )
+
+        instance_app = ConfidentialClientApplication(
+            client_id=agent_app_instance_id,
+            authority=authority,
+            client_credential={"client_assertion": agent_token_result},
+        )
+
+        agent_instance_token = instance_app.acquire_token_for_client(
+            ["api://AzureAdTokenExchange/.default"]
+        )
+
+        assert agent_instance_token
+        assert agent_token_result
+
+        # future scenario where we don't know the blueprint id upfront
+
+        token = agent_instance_token.get("access_token")
+        if not token:
+            logger.error("Failed to acquire agentic instance token, %s", agent_instance_token)
+            raise ValueError(f"Failed to acquire token. {str(agent_instance_token)}")
+
+        payload = jwt.decode(token, options={"verify_signature": False})
+        agentic_blueprint_id = payload.get("xms_par_app_azp")
+        logger.debug("Agentic blueprint id: %s", agentic_blueprint_id)
+
+        return agent_instance_token["access_token"], agent_token_result
+
+    async def get_agentic_user_token(
+        self, agent_app_instance_id: str, upn: str, scopes: list[str]
+    ) -> Optional[str]:
+        """Gets the agentic user token for the given agent application instance ID and user principal name and the scopes.
+
+        :param agent_app_instance_id: The agent application instance ID.
+        :type agent_app_instance_id: str
+        :param upn: The user principal name.
+        :type upn: str
+        :param scopes: The scopes to request for the token.
+        :type scopes: list[str]
+        :return: The agentic user token, or None if not found.
+        :rtype: Optional[str]
+        """
+        breakpoint()
+        if not agent_app_instance_id or not upn:
+            raise ValueError(
+                "Agent application instance Id and user principal name must be provided."
+            )
+
+        instance_token, agent_token = await self.get_agentic_instance_token(
+            agent_app_instance_id
+        )
+
```

Comment:
consider a check here and logging if there's an error come back. If these didn't create right, we'll get unexpected errors below.